gsoc-2023/ideas.md

```diff
@@ -0,0 +1,263 @@
+
+#### Reading & Related Material
+
+ * https://unikraft.org/docs/features/security/
+ * https://dl.acm.org/doi/abs/10.1145/3503222.3507759
+ * https://github.com/a13xp0p0v/linux-kernel-defence-map
+
+---
+
+### MSI/MSI-X Interrupt Driver Support (x86, KVM)
+
+| | |
+|-|-|
+| **Difficulty** | 3/5 |
+| **Project Size** | Variable (175 or 350 hours) |
+| **Maximum instances** | 1 |
+| **Constraints/requirements** | C programming skills, OS/driver concepts, familiarity with x86 architecture. Experience with interrupt controllers, the PCI standard, and virtio is a plus. |
+
+#### Description
+
+Unikraft's interrupt system on KVM/x86 is based on the legacy 8259 programmable interrupt controller (PIC) which is limited to 15 interrupt lines and programmed using slow port I/O.
+Because the interrupt vector space is so small, when having many devices, interrupt lines need to be shared.
+The operating system thus has to query multiple drivers to figure out which one is responsible for handling an interrupt on a shared line and which device caused the interruption. This entails reading many device registers, leading to high interrupt latency.
+However, in order to achieve high I/O performance, one key is to keep the software overhead of interrupt handlers minimal.
+Modern devices (including virtio) even reduce software processing time further by setting up multiple interrupt lines, like one for each I/O queue they provide.
+With having just 15 interrupt lines, line sharing happens even with only a few devices attached.
+Advanced programmable interrupt controllers (APICs) utilizing the MSI and MSI-X PCI standards overcome this problem by providing a much larger interrupt vector space (currently up to 2048). Interrupt line sharing is thus not required anymore.
+
+In this project, you will (1) develop a driver for interfacing a modern x86 I/O APIC and extend Unikraft’s libpci to support MSI/MSI-X, and (2) enable the existing virtio drivers to make use of these technologies.
+
+#### Reading & Related Material
+
+ * [Intel 64 and IA-32 Architectures Software Developer's Manual](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html)
+ * https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html
+ * https://wiki.osdev.org/Interrupts
+ * https://wiki.osdev.org/APIC
+ * https://wiki.osdev.org/PCI#Message_Signaled_Interrupts
+ * https://habr.com/en/post/501912/
+
+---
+
+### Unit Tests in Internal and External Libraries via uktest
+
+| | |
+|-|-|
+| **Difficulty** | 2/5 |
+| **Project Size** | Variable (175 or 350 hours) |
+| **Maximum instances** | 1 |
+| **Constraints/requirements** | C programming skills, experience with unit testing is a plus. |
+
+#### Description
+
+Since release 0.6, Unikraft provides the [`uktest` framework](https://github.com/unikraft/unikraft/tree/staging/lib/uktest) as an internal library to crate unit tests for core and external libraries.
```

Comment:
```suggestion
Since release 0.6, Unikraft provides the [`uktest` framework](https://github.com/unikraft/unikraft/tree/staging/lib/uktest) as an internal library to create unit tests for core and external libraries.
```